Sources/ConfigCat/ConfigCatClient.swift

```diff
@@ -37,177 +37,232 @@ public final class ConfigCatClient : NSObject, ConfigCatClientProtocol {
+            configService = ConfigService(log: log, fetcher: fetcher, cache: configCache, pollingMode: mode, sdkKey: sdkKey)
         }
     }
 
     deinit {
         ConfigCatClient.sdkKeys.remove(sdkKey)
     }
 
-    func getSettingsAsync() -> AsyncResult<[String: Any]> {
-        if let overrideDataSource = self.overrideDataSource, overrideDataSource.behaviour == .localOnly {
-            return AsyncResult<[String: Any]>.completed(result: overrideDataSource.getOverrides())
+    // MARK: ConfigCatClientProtocol
+
+    public func getValue<Value>(for key: String, defaultValue: Value, user: ConfigCatUser? = nil, completion: @escaping (Value) -> ()) {
+        if key.isEmpty {
+            assert(false, "key cannot be empty")
+        }
+        getSettings { settings in
+            completion(self.getValueFromSettings(settings: settings, key: key, defaultValue: defaultValue, user: user))
+        }
+    }
+
+    @objc public func getAllKeys(completion: @escaping ([String]) -> ()) {
+        getSettings { settings in
+            completion([String](settings.keys))
+        }
+    }
+
+    @objc public func getVariationId(for key: String, defaultVariationId: String?, user: ConfigCatUser? = nil, completion: @escaping (String?) -> ()) {
+        if key.isEmpty {
+            assert(false, "key cannot be empty")
+        }
+        getSettings { settings in
+            completion(self.getVariationIdFromSettings(settings: settings, key: key, defaultVariationId: defaultVariationId, user: user))
+        }
+    }
+
+    @objc public func getAllVariationIds(user: ConfigCatUser? = nil, completion: @escaping ([String]) -> ()) {
+        getSettings { settings in
+            completion(self.getAllVariationIdsFromSettings(settings: settings, user: user))
+        }
+    }
+
+    @objc public func getKeyAndValue(for variationId: String, completion: @escaping (KeyValue?) -> ()) {
+        getSettings { settings in
+            completion(self.getKeyAndValueFromSettings(settings: settings, variationId: variationId))
         }
+    }
 
-        guard let refreshPolicy = self.refreshPolicy else {
-            return AsyncResult<[String: Any]>.completed(result:[:])
+    @objc public func getAllValues(user: ConfigCatUser? = nil, completion: @escaping ([String: Any]) -> ()) {
+        getSettings { settings in
+            completion(self.getAllValuesFromSettings(settings: settings, user: user))
         }
+    }
 
-        if let overrideDataSource = self.overrideDataSource {
+    @objc public func refresh(completion: @escaping () -> ()) {
+        if let configService = configService {
+            configService.refresh(completion: completion)
+        } else {
+            log.warning(message: "The ConfigCat SDK is in local-only mode. Calling .refresh() has no effect.")
+            completion()
+        }
+    }
```

Comment:
In the event of refreshing frequently, is there a mechanism that ensures there is only one request in flight at the same time?

Reply:
After digging a bit, it seems it's handled here, if I understand correctly (`ConfigService:113`).

```swift
private func fetchIfOlder(time: Date, preferCache: Bool = false, completion: @escaping (Config) -> Void) {
        mutex.lock()
        defer {
            mutex.unlock()
        }
        if cachedEntry.isEmpty || cachedEntry.fetchTime > time {
            let json = readConfigCache()
            if !json.isEmpty && json != cachedEntry.jsonString {
                let parseResult = json.parseConfigFromJson()
                switch parseResult {
                case .success(let config):
                    cachedEntry = ConfigEntry(jsonString: json, config: config, eTag: "")
                case .failure(let error):
                    log.error(message: "An error occurred during JSON deserialization. %@", error.localizedDescription)
                }
            }
            if cachedEntry.fetchTime > time {
                completion(cachedEntry.config)
                return
            }
        }
        if preferCache && initialized.get() {
            completion(cachedEntry.config)
            return
        }
        // There's an ongoing fetch, save the callback to call it later when the fetch succeeds.
        if completions != nil {
            completions?.enqueue(item: completion)
            return
        }
        // No fetch is running, initiate a new one.
        completions = MutableQueue<(Config) -> Void>()
        completions?.enqueue(item: completion)
        fetcher.fetch(eTag: cachedEntry.eTag) { response in
            self.processResponse(response: response)
        }
    }
```

Reply:
Yep, that is the part which ensures that only a single fetch is running at one time. I put a bit more detailed comments to that section to make the functionality clearer.